### ISSUE TITLE
Compiled code contain jsx code.

### ISSUE THREAD

**Issue by IvanovRoman** [microsoft/TypeScript#54411]
# Bug Report

Compiled code contain `jsx` code `<Label label='aaaa'/>`

### 🕗 Version & Regression Information

Typescript version 5.1.0-beta

### ⏯ Playground Link

[Playground link with problem code](https://www.typescriptlang.org/play?declaration=false&target=2&module=2&ts=5.1.0-beta#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQduEAdqvADLIBGWANgFyYqMAHQAxAMIAeAN5xe3PoLZRgzAOY04AXwB8cALxwwxMOgP6pCnr10yTEMyOt9tUgPQvb9OmphYoAjwsOAAhNQATAAVTVCkAFUMUZgBPABo4AGk4LAAPf2YI9ABrLBSIAjhEo2RU-Rk6ODguAHkuACssfEF4+iauTLLBTL7mzhsAfmUYVQ04AB8hPFFsZYA5CAisem0GAgBXZnxgFmbIhIzsvIKiuFLyyvjdAApWjq6YHoyBoazv8b4UzgKjU6gWS3wIlW+A2WwAlIJwoUYo44vFLvVGnBKDB9lBmHA5G9OvhvoN0mNFLwdDs9odjqcAO7AGAACyRESkURy+SwhXQEHepLgADUeTcSmUKnAohkJOK+bd7tKoi8sRISJBmHzPhDRBrwCwdfEUmAsFzdGksWBgLhioJngA3ZC8fZYQRRADaIoAuhkWBJWbV1O64E6XW6Pd6fXDDPpHRBgBFYxYZbbilyMmLFhJdHRYw0mk0cXiCbgFKh0NdFehofrNUbmDApC0QCzM6Lwbm4AAyMKRFFmTFFkfGVTO-xwANBjQhIzPFPD0dNbQ+ZfFxUBBeErHrotMVjwIltEkwMllf5UjIiG+UNg6JJs4CoEQOMyjPdNA-3t-mQlwG9X3TV4Tw+T0fhSP04CfF9p2DLA4WvW8sHvbQUHQKIP0-b94D4LAQB1JIpANLVCJkQDfzQ9xdCwvdgEqV4AWpAwWLgQ4tgINQsGTHdP3XEt8RyXh8J1Wj112Piml3T8BIJZ5pMkuApDzRTlxkGBTSwaUuCYwxWLIEENDIOAJiU3SvAMGQdKpKj9EEaybAk1SRxkPCCKbJznI8FTJLhMSV2k3Zdl8JsAiCXAQnia4AElmDAfYYEHdBCyLZ1XVDQzNGkuDZ0EOsREDeCAFFHR1AAJWoImEqApDK+IAFl2Fi+KYCK4T3JgGi6GCnCqhiuKEry4RxGkKL8mahKkv0IxfzjJS1Ba6DNIMMh-HyYzyJvSi4Gonw8kgWA4F60IxpgCb4CMZk2Q5BJ+papKMjINK3TIR6cpDMgXlO86MnDdL-WYQrZ0XXjsSwXFBLkZ6sABoGQ0EOd9HerBniwERcDxSgm3iZAoBDURodjVdur8hhesFdokjkNbdTIYziakE67oSwlAK4SJngpx6adesyqXkKkVuQYXkGM6i4VsmggA)

[Playground link with work code](https://www.typescriptlang.org/play?declaration=false&target=2&module=2&ts=5.0.4#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQduEAdqvADLIBGWANgFyYqMAHQAxAMIAeAN5xe3PoLZRgzAOY04AXwB8cALxwwxMOgP6pCnr10yTEMyOt9tUgPQvb9OmphYoAjwsOAAhNQATAAVTVCkAFUMUZgBPABo4AGk4LAAPf2YI9ABrLBSIAjhEo2RU-Rk6ODguAHkuACssfEF4+iauTLLBTL7mzhsAfmUYVQ04AB8hPFFsZYA5CAisem0GAgBXZnxgFmbIhIzsvIKiuFLyyvjdAApWjq6YHoyBoazv8b4UzgKjU6gWS3wIlW+A2WwAlIJwoUYo44vFLvVGnBKDB9lBmHA5G9OvhvoN0mNFLwdDs9odjqcAO7AGAACyRESkURy+SwhXQEHepLgADUeTcSmUKnAohkJOK+bd7tKoi8sRISJBmHzPhDRBrwCwdfEUmAsFzdGksWBgLhioJngA3ZC8fZYQRRADaIoAuhkWBJWbV1O64E6XW6Pd6fXDDPpHRBgBFYxYZbbilyMmLFhJdHRYw0mk0cXiCbgFKh0NdFehofrNUbmDApC0QCzM6Lwbm4AAyMKRFFmTFFkfGVTO-xwANBjQhIzPFPD0dNbQ+ZfFxUBBeErHrotMVjwIltEkwMllf5UjIiG+UNg6JJs4CoEQOMyjPdNA-3t-mQlwG9X3TV4Tw+T0fhSP04CfF9p2DLA4WvW8sHvbQUHQKIP0-b94D4LAQB1JIpANLVCJkQDfzQ9xdCwvdgEqV4AWpAwWLgQ4tgINQsGTHdP3XEt8RyXh8J1Wj112Piml3T8BIJZ5pMkuApDzRTlxkGBTSwaUuCYwxWLIEENDIOAJiU3SvAMGQdKpKj9EEaybAk1SRxkPCCKbJznI8FTJLhMSV2k3Zdl8JsAiCXAQnia4AElmDAfYYEHdBCyLZ1XVDQzNGkuDZ0EOsREDeCAFFHR1AAJWoImEqApDK+IAFl2Fi+KYCK4T3JgGi6GCnCqhiuKEry4RxGkKL8mahKkv0IxfzjJS1Ba6DNIMMh-HyYzyJvSi4Gonw8kgWA4F60IxpgCb4CMZk2Q5BJ+papKMjINK3TIR6cpDMgXlO86MnDdL-WYQrZ0XXjsSwXFBLkZ6sABoGQ0EOd9HerBniwERcDxSgm3iZAoBDURodjVdur8hhesFdokjkNbdTIYziakE67oSwlAK4SJngpx6adesyqXkKkVuQYXkGM6i4VsmggA)

### 💻 Code
```ts
import * as React from 'react';

const Label: React.FC<{ label: string; }> = props => <label>{props.label}</label>;

interface BindProps<T = any, K extends keyof T = any> {
  bObject: T;
  bKey: K;
  bLabel?: string | React.ReactNode;
}

function bind<T, K extends keyof T>(bObject: T, bKey: K, bLabel?: string | React.ReactNode): BindProps<T, K> {
  return { bObject, bKey, bLabel };
}

function withBind<P extends object, V extends keyof P, C extends keyof P>(
  Component: React.ComponentType<P>,
  pick: (value: P[V], onChange: (value: P[V]) => void) => Pick<P, V | C>
) {
    return class extends React.Component<Omit<P, V | C> & BindProps> {
        private onChange = () => {
        };

        render() {
            const { bObject, bKey, bLabel, ...rest } = this.props;
            const props = { ...pick(bObject[bKey], this.onChange), ...rest } as P;
            const element = <Component {...props} />;
            if (bLabel === undefined) {
                return element;
            }
            
            return (
                <>
                  {typeof bLabel === 'string' ? <Label label={bLabel} /> : bLabel}
                  {element}
                </>
            );
        }
    }
}

interface TextInputProps {
    value: string;
    onChange: React.ChangeEventHandler<HTMLInputElement>;
}

const TextInput: React.FC<TextInputProps> = props => <input type='text' {...props} />;

export const BTextInput = withBind<TextInputProps, 'value', 'onChange'>(TextInput, (value, onChange) => {
  return { value, onChange: e => onChange(e.currentTarget.value) };
});

const obj = { text: '' };
<BTextInput {...bind(obj, 'text', <Label label='aaaa' />)} />;
```

### 🙁 Actual behavior

```ts
React.createElement(exports.BTextInput, Object.assign({}, bind(obj, 'text', <Label label='aaaa'/>)));
```

### 🙂 Expected behavior

```ts
React.createElement(exports.BTextInput, Object.assign({}, bind(obj, 'text', React.createElement(Label, { label: 'aaaa' }))));
```

**Comment by RyanCavanaugh** [comment]
Simpler:
```ts
// Not transformed
const n1 = <div {...<span />} />;
// Transformed
const n2 = <div {...(() => {})} />;
```
This is specific to when we need to call `Object.assign` or `__assign` to emit the spread. When the emit uses a native JS spread, then we don't see this

**Comment by ericrafalovsky** [comment]
Once merged, can the fix be cherry picked into a 5.1.x release? This seems like a regression currently in 5.1.2-rc.

**Comment by thatsmydoing** [comment]
There's a minor variation of this that's not fixed in 5.1.6 or the latest nightly

```
<div {...{ prop: <span /> }} />
```

**Comment by RyanCavanaugh** [comment]
Fixed the variation in the linked issue/PR
